## HW1

胡志峰 16307130177

### Problem 1

The output of the frequent words are AAC, which appears 3 times in the DNA sequence.

Suppose the nucleotides are uniformly distributed in DNA sequences and we don't count for overlaps, then the probability of any particular 3-mers to occur 3 times in a DNA sequence of length 19 is approximately
$$
\left(\frac{1}{4^3}\right)^3\times{19-3\choose3}\times \left(1-\frac{1}{4^3}\right)^{19-3}\times {4^3}\approx 0.106
$$
which means it is particularly frequent.



### Problem 2

It is not correct, because Q is mutated during each generation. We can preserve Q before making any modify to it to prevent mutation of the original string Q.

Formally, we create a new string P which is the same as Q, and $P[j] \leftarrow A[k]$ each time we generate a new string.

### Problem 3

the pseudo code of the algorithm is as follow: 

```
For each k-mer Q in the input DNA sequence Text:
	P <- reverse and complement of Q
	if P and Q are the same:
		output Q
```

In python, we can calculate the reverse complement by

```python
P = ''.join([{'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}[n] for n in Q[::-1]])
```

### Problem 4

As we know, the replication starts from the oriC. Recall that during the replication, the reversed strand will be a single-stranded DNA, which means the Cytosine (C) on it has a high rate mutating into thymine (T) through deamination, hence if we walk along the result of the replication in the forward direction (5' to 3') , and count for the difference of the number of G and C,  it first increasing and then decreasing, and the lowest point is the oriC. Since circular genomes can break at any point, the oriC can occasionally be the point where we break the genomes. In this diagram, the lowest point is at that the end position, so the oriC is located at either the beginning of the string of nucleotides or at the end of it.

### Problem 5

The profile can be built by calculating the probability of each nucleotides in each column, the calculation yields

```
A: 0.0 0.5 0.5 0.3 0.1 0.3 1.0 0.8 0.0 0.2
C: 0.5 0.0 0.4 0.4 0.0 0.0 0.0 0.0 0.0 0.2
G: 0.2 0.5 0.1 0.1 0.2 0.6 0.0 0.1 0.0 0.0
T: 0.3 0.0 0.0 0.2 0.7 0.1 0.0 0.1 1.0 0.6
```

$$
Pr(S_0|\mathbb {profile})=0.5\times 0.5\times 0.5\times 0.3\times 0.7\times 0.1\times 1.0\times 0.1\times 1.0\times 0.6\approx0.0001575
$$

### Problem 6

Incorrect. The Median String Problem wants a k-mer pattern among all k-mer patterns that minimizing the score, the algorithm proposed by Mr.Fuzzy can only find the k-mer that is occurred in the t DNA string.

### Problem 7

To verify which result can be more likely to be the motifs of NobX, we can calculate the score of each result. The score of a set of motifs as taught in class is the sum of distance of the consensus between each motif.

In the first set, the consensus is TTACCTTAAC, hence the score is 2 + 1 + 1 + 1 + 2 + 1 + 2 + 1 + 1 + 1 = 13.

in the second set, with consensus being AA(C)ACCGGCAC (the character in the parenthesis means that it can replace the character before it, and the string is still a consensus, although this will not affect the score), the score is 3 + 3 + 2 + 2 + 2 + 2 + 1 + 2 + 2 + 0 = 19.

The first set has a lower score than the second, which means the consensus of the first answer is more likely to be the binding sites than the second one, hence the second answer is more likely wrong.